device.go

```diff
@@ -229,37 +230,182 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	fieldLogger := agentLog.WithField("host-dev", device.ContainerPath)
+
+	var group string
+
+	for _, opt := range device.Options {
+		tokens := strings.Split(opt, "=")
+		if len(tokens) != 2 {
+			return fmt.Errorf("Malformed VFIO option \"%s\"", opt)
+		}
+
+		hostBdf := tokens[0]
+		guestPCIPath := tokens[1]
+
+		guestAddr, err := getDevicePCIAddress(guestPCIPath)
+		if err != nil {
+			return err
+		}
+
+		tokens = strings.Split(guestAddr, "/")
+		guestBdf := tokens[len(tokens)-1]
+
+		s.vfioDeviceMap[hostBdf] = guestBdf
+
+		err = vfioRebindDevice(guestBdf)
+		if err != nil {
+			return err
+		}
+
+		groupPath := filepath.Join(sysBusPrefix, guestBdf, "iommu_group")
+		groupLink, err := os.Readlink(groupPath)
+		if err != nil {
+			return fmt.Errorf("Failed to read %s: %s", groupPath, err)
+		}
+
+		if group == "" {
+			group = filepath.Base(groupLink)
+		} else if group != filepath.Base(groupLink) {
+			// If PCI devices associated with the same
+			// VFIO device (and therefore group) in the
+			// host don't end up in the same group in the
+			// guest, something has gone horribly wrong
+			return fmt.Errorf("%s is not in guest IOMMU group %s", guestBdf, group)
+		}
+		fieldLogger.WithField("host-bdf", hostBdf).WithField("guest-bdf", guestBdf).Debug("VFIO: Device complete")
+	}
+
+	vmpath, err := getDeviceName(s, filepath.Join("vfio", group))
+	if err != nil {
+		return fmt.Errorf("Couldn't find device for guest VFIO group %s: %s",
+			group, err)
+	}
+	device.VmPath = vmpath
+
+	fieldLogger.WithField("guest-dev", device.VmPath).Debug("VFIO: Group complete")
+
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.VmPath)
+}
+
+func VfioTranslateEnv(s *sandbox, env []string) []string {
+	newEnv := make([]string, 0, len(env))
+
+	for _, e := range env {
+		if strings.HasPrefix(e, "PCIDEVICE_") {
+			for host, guest := range s.vfioDeviceMap {
+				e = strings.ReplaceAll(e, host, guest)
+			}
+		}
+		newEnv = append(newEnv, e)
+	}
+
+	return newEnv
+}
+
+// updateSpecDevice updates a device list in the OCI spec to make it
+// include details appropriate for the VM, instead of the host.  It is
+// given the host path to the device (to locate the device in the
+// original OCI spec), the VM path which it uses to determine the VM
+// major/minor number, and the final path with which to present the
+// device in the (inner) container
+func updateSpecDevice(spec *pb.Spec, devIdx devIndex, hostPath, vmPath, finalPath string) error {
+	// If no hostPath is provided, we won't be able to match and
+	// update the device in the OCI spec device list. This is an
+	// error.
+	if hostPath == "" {
```

Comment:
Should there be a check for empty `vmPath` as well?

Reply:
If it's empty, it will fail on the `syscall.Stat()` below with a somewhat sensible error message, so I would not create a special case for that.